dask/array/core.py

```diff
@@ -3493,6 +3498,72 @@ def _vindex(x, *indexes):
     return result_1d.reshape(broadcast_shape + result_1d.shape[1:])
 
 
+def _vindex_dask_arrays(x, *indices):
+    """Point wise indexing with Dask Arrays"""
+
+    from .creation import arange
+
+    x = asarray(x)
+
+    oth_inds = []
+    arr_inds = []
+    out_axes = []
+    for i, ind in enumerate(indices):
+        if isinstance(ind, Array):
+            oth_inds.append(slice(None))
+            arr_inds.append(asarray(ind))
+            if 0 not in out_axes:
+                out_axes.append(0)
+        elif isinstance(ind, Integral):
+            oth_inds.append(ind)
+        else:
+            oth_inds.append(ind)
+            arr_inds.append(slice(None))
+            out_axes.append(len(arr_inds))
+    oth_inds = tuple(oth_inds)
+    arr_inds = tuple(arr_inds)
+    out_axes = tuple(out_axes)
+
+    if not all(o == slice(None) for o in oth_inds):
+        x = x.vindex[oth_inds]
+
+    x_inds = tuple(
+        arange(s, chunks=c) for s, c in zip(x.shape, x.chunks)
+    )
+
+    idx_vals = atop(
+        _vindex_dask_arrays_chunk,
+        out_axes,
+        x, tuple(range(1, 1 + x.ndim)),
+        *concat([
+            (e_0, (0,), e_i, (i,)) if isinstance(e_0, Array) else
+            (e_0, None, e_i, (i,))
+            for i, (e_0, e_i) in enumerate(zip(arr_inds, x_inds), start=1)
+        ]),
+        dtype=x.dtype,
+        concatenate=True
+    )
+
+    return idx_vals
+
+
+def _vindex_dask_arrays_chunk(xc, *args):
+    inds = args[::2]
+    x_inds = args[1::2]
+
+    xc_inds = []
+    for i, (e_0, e_i) in enumerate(zip(inds, x_inds)):
+        if isinstance(e_0, np.ndarray):
+            xc_inds.append(np.equal.outer(e_0, e_i).nonzero()[1])
```

Comment:
I think `np.in1d` could be much faster here (O(n log n) vs O(n^2)). This requires allocating an array for the full outer product of the indices.

Reply:
Sure, that makes sense.

When I previously was playing with this in the gist, it used only Dask Array operations without `atop`. So when pushing it all into `atop`, expect that there remains some things from the old approach that are suboptimal.

Happy to fix any that come up.

Reply:
Looking back at it, I think the current code was chosen to preserve ordering. For example...

```python
>>> import numpy as np
>>> e_0 = np.array([23, 21, 24])
>>> e_i = np.arange(21, 25)
>>> np.equal.outer(e_0, e_i).nonzero()[1]
array([2, 0, 3])
>>> np.in1d(e_i, e_0).nonzero()[0]
array([0, 2, 3])
```

Not sure if there is an obvious way to get this from `in1d` or a similarly performant function, but would be happy to try. Expect this would be a good point to improve.

Reply:
If we're OK with a pandas dependency (which we would probably rather avoid in dask.array), this version does the lookup in linear time:
```
In [14]: import pandas as pd

In [15]: pd.Index(e_i).get_indexer(e_0)
Out[15]: array([2, 0, 3])
```

It looks like this could also be done with `np.searchsorted()`:
```
In [24]: np.searchsorted(np.sort(e_i), e_0)
Out[24]: array([2, 0, 3])
```
(you should add tests/think carefully about how to properly handle repeated indices)

Reply:
Thanks for the tip. Will give that a look.

So we do have tests for repeated indices in this PR. Please let me know if we are missing something in particular there.

Should add we don't have tests for non-sequential indices, which we need. Happy to add those.

Reply:
FWICT `searchsorted` works well. We can skip the `sort` as `e_i` is always sorted. Updated the PR accordingly. Thanks again for the suggestion.